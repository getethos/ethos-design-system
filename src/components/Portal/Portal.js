import { useEffect, useRef, useState } from 'react'
import { createPortal } from 'react-dom'
import PropTypes from 'prop-types'

/**
 * Hooks handles the creation and tear down of the portal's root
 * element.
 *
 * @private
 *
 * @param {string} [id] - The optional id of the target container. If the `id` is not passed,
 * the target container will be a div without an `id`
 *
 * @return {HTMLElement} - The DOM node to use as the Portal target
 */
function usePortal(id) {
  const [isAttached, setIsAttached] = useState(false)
  const elRef = useRef(
    document.getElementById(id) || document.createElement('div')
  )

  useEffect(() => {
    const el = elRef.current
    setIsAttached(!!el.parentElement)

    if (!isAttached) {
      el.id = id
      document.body.appendChild(el)
      setIsAttached(true)
    }

    return () => {
      if (isAttached) {
        elRef.current.parentElement.removeChild(el)
      }
    }
  }, [id, elRef, isAttached])

  return elRef.current
}

/**
 * Dynamically create's a portal component that appends to the end of the DOM
 *
 * @public
 *
 * @param {object} props - the component's props
 * @param {string} [props.id] - The optional id of the target container. If the `id` is not passed,
 * the target container will be a div without an `id`
 *
 * @param {React.children} props.children - the children element
 *
 * @example ```
 * <Portal id="glados">
 *   <h1>This was a triumph</h1>
 * </Portal>
 * ```
 *
 * @return {React.ReactPortal}
 */
export const Portal = ({ id, children }) => {
  const root = usePortal(id)

  return createPortal(children, root)
}

Portal.propTypes = {
  id: PropTypes.string,
  children: PropTypes.node,
}
