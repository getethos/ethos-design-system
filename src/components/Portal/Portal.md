## Description
The `Portal` element provides a way to create React portals dynamically. The component will create a new `DOM` node that attaches to the end of the `document.body` and render's the passed children within that DOM node.

## Example

```jsx
const styles = {
    position: 'fixed',
    top: '20px',
    right: '20px',
    border: '1px solid var(--BrandForest)',
    color: 'var(--BrandForest)',
    background: 'var(--BrandMoss)',
    borderRadius: '3px',
    padding: 'var(--Space-16) var(--Space-24)',
};
<Portal id="foo">
    <div style={styles}>
        <h1>Diagon Alley!</h1>
    </div>
</Portal>
```