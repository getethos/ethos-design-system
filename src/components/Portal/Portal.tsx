import { useEffect, useRef, useState } from 'react'
import { createPortal } from 'react-dom'
/**
 * Hooks handles the creation and tear down of the portal's root
 * element.
 *
 * @private
 *
 * @param {string} [id] - The optional id of the target container. If the `id` is not passed,
 * the target container will be a div without an `id`
 *
 * @return {HTMLElement} - The DOM node to use as the Portal target
 */
function usePortal(id) {
  const [isAttached, setIsAttached] = useState(false)
  const elRef = useRef(
    document.getElementById(id) || document.createElement('div')
  )
  useEffect(() => {
    const el = elRef.current
    setIsAttached(!!el.parentElement)
    if (!isAttached) {
      el.id = id
      document.body.appendChild(el)
      setIsAttached(true)
    }
    return () => {
      if (elRef.current.parentElement && isAttached) {
        elRef.current.parentElement.removeChild(el)
      }
    }
  }, [id, elRef, isAttached])
  return elRef.current
}
type PortalProps = {
  id?: string
}
/**
 * Dynamically create's a portal component that appends to the end of the DOM
 *
 * @public
 *
 * @param {object} props - the component's props
 * @param {string} [props.id] - The optional id of the target container. If the `id` is not passed,
 * the target container will be a div without an `id`
 *
 * @param {React.children} props.children - the children element
 *
 * @example ```
 * <Portal id="glados">
 *   <h1>This was a triumph</h1>
 * </Portal>
 * ```
 *
 * @return {React.ReactPortal}
 */
export const Portal: React.SFC<PortalProps> = ({ id, children }) => {
  if (typeof document === 'undefined' || !document) {
    return null
  }
  const root = usePortal(id)
  return createPortal(children, root)
}
