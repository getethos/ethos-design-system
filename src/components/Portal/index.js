export { Portal } from './Portal.js'
