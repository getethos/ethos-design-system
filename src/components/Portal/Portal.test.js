import React from 'react'
import { Portal } from './Portal'
import renderer from 'react-test-renderer'
import ReactDOM from 'react-dom'

const TEST_ID = 'my-portal'

describe('Portal', () => {
  beforeAll(() => {
    // createPortal needs to be mocked
    // SEE: https://github.com/facebook/react/issues/11565
    ReactDOM.createPortal = jest.fn((element) => {
      return element
    })
  })

  afterEach(() => {
    ReactDOM.createPortal.mockClear()
  })

  it("render's a portal", () => {
    const tree = renderer
      .create(
        <Portal id={TEST_ID}>
          <h1>hello world</h1>
        </Portal>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
