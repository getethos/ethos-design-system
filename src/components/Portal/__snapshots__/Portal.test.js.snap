// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Portal render's a portal 1`] = `
<h1>
  hello world
</h1>
`;
